[Rando] Clock Shuffle + Logic

mckinlee opened the pull request:
Turns the 3-day cycle into collectible items that you need to find before accessing different time periods.

**How it works:**

- Each half-day (Day 1, Night 1, Day 2, Night 2, Day 3, Night 3) becomes a randomized item
- You can only visit time periods you've collected
- If you try to access a locked time period, you'll be redirected to the earliest one you own
- If you don't own any half-days, you'll be sent to the start of the final hours

**Settings:**

- Random: Half-days are shuffled randomly
- Ascending: Unlock in order (Day 1 → Night 1 → Day 2, etc.)
- Descending: Unlock in reverse order (Night 3 → Day 3 → Night 2, etc.)

## Must Have
- [x] ~~Logic, I think is fully implemented, but could use a few passes~~
- [x] ~~Logic needed for pots, crates, snowballs, you name it.~~
- [ ] Needs a thorough review and testing

## Nice to Have
- [x] ~~Animated clock item (Cal will work on this, thank you!)~~
- [x] ~~Make all "remaining hours" calculation based on obtained clock items~~
- [x] ~~Add daytelops (This is complete on my local machine, gonna test more before pushing)~~
- [x] ~~Configurable final hours length (Default is 6 hours before moon crash)~~
- [x] ~~SoDT textbox to state the correct half-day~~
- [x] ~~BetterSoDT feedback~~

## Needs Fixing
- [x] ~~SkipSoTCutscenes conflicts causing wierd cascading miscalculated overcorrection~~
- [x] ~~"Queue Randomizer Item Gives" section of save editor should not be conditional on RO_CLOCK_SHUFFLE~~
- [x] ~~On file load, should not correct for owl saves~~
- [x] ~~Deku hopping during a correction causes infinite loop of sinking~~
- [x] ~~seed generation issues for french vanilla and glitchless when clock shuffle is disabled~~
- [x] ~~SkipSoTCutscene still conflicts on N3 and reverts back to start of N3~~


<!--- section:artifacts:start -->
### Build Artifacts
  - [2ship-windows.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/5030047023.zip)
  - [2ship-linux.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/5030104177.zip)
  - [2ship-mac.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/5030923191.zip)
<!--- section:artifacts:end -->

mckinlee committed:
Add Clocks as Items feature

mckinlee committed:
refactor

mckinlee committed:
Merge remote-tracking branch 'upstream/develop' into clocks-as-items

mckinlee committed:
some clean up

mckinlee committed:
forgot to clangify

mckinlee committed:
use proper randoinf flags and some fixes

mckinlee committed:
proper skipsotcutscene fix

Caladius committed:
Animated Clock Item (Thank you Caladius!)

* Close attempt at animating the clock, not perfect but a good start.

* Git gud Clock

mckinlee committed:
fix clock item get crash

Co-Authored-By: Caladius <Caladius@users.noreply.github.com>

mckinlee committed:
some enhancements

Eblo commented on mm/2s2h/DeveloperTools/SaveEditor.cpp:
```diff
@@ -885,7 +886,74 @@ void DrawItemsAndMasksTab() {
     }
     UIWidgets::Checkbox("Safe Mode", &safeMode);
 
+    if (gSaveContext.save.shipSaveInfo.saveType == SAVETYPE_RANDO &&
```
Since the next chunk is also within a condition for the rando save type, you could just move this to the start of that, with only the clock shuffle condition.

Eblo commented on mm/2s2h/DeveloperTools/SaveEditor.cpp:
```diff
@@ -885,7 +886,74 @@ void DrawItemsAndMasksTab() {
     UIWidgets::Checkbox("Safe Mode", &safeMode);
 
+    if (gSaveContext.save.shipSaveInfo.saveType == SAVETYPE_RANDO &&
+        CVarGetInteger("gRando.Options.RO_CLOCK_SHUFFLE", 0)) {
```
```suggestion
        RANDO_SAVE_OPTIONS[RO_CLOCK_SHUFFLE]) {
```

Eblo commented on mm/2s2h/Enhancements/Songs/SkipSoTCutscenes.cpp:
```diff
@@ -26,8 +27,26 @@ void RegisterSkipSoTCutscenes() {
-            gSaveContext.save.time = CLOCK_TIME(6, 0) - 1;
+
+            // Set time appropriately if clock shuffle is enabled
+            if (CVarGetInteger("gRando.Options.RO_CLOCK_SHUFFLE", 0)) {
```
```suggestion
            if (IS_RANDO && RANDO_SAVE_OPTIONS[RO_CLOCK_SHUFFLE]) {
```

Eblo commented on mm/2s2h/Rando/CheckTracker/CheckTracker.cpp:
```diff
@@ -514,6 +520,40 @@ void CheckTrackerWindow::Draw() {
     ImGui::PopStyleVar(1);
+
+    // Compact Clocks-as-Items indicator row
+    if (RANDO_SAVE_OPTIONS[RO_CLOCK_SHUFFLE] && CVAR_CLOCK_SEGMENTS_SHOW) {
```
Clocks should be tracked by the item tracker, not the check tracker. The check tracker details checks, i.e. where the items come from, while the item tracker details the collected items themselves. In this case, we care about displaying which days/nights are available, ergo item tracker. Triforce Hunt is a good example of a rando setting-specific item being tracked. If we can decide on a good icon for it, I think a simple text overlay, like with the skulltula tokens, would do fine enough.

Eblo commented on mm/2s2h/Rando/Logic/Regions/Central.cpp:
```suggestion
CHECK(RC_CLOCK_TOWN_NORTH_TINGLE_MAP_01, IS_DAY && CAN_USE_PROJECTILE && CAN_AFFORD(RC_CLOCK_TOWN_NORTH_TINGLE_MAP_01)),
CHECK(RC_CLOCK_TOWN_NORTH_TINGLE_MAP_02, IS_DAY && CAN_USE_PROJECTILE && CAN_AFFORD(RC_CLOCK_TOWN_NORTH_TINGLE_MAP_02)),
```

Eblo commented on mm/2s2h/Rando/Logic/Regions/Central.cpp:
```diff
@@ -199,19 +227,40 @@ static RegisterShipInitFunc initFunc([]() {
-            EXIT(ENTRANCE(CURIOSITY_SHOP, 0),               ENTRANCE(WEST_CLOCK_TOWN, 4), true),
-            EXIT(ENTRANCE(TRADING_POST, 0),                 ENTRANCE(WEST_CLOCK_TOWN, 5), true),
+            EXIT(ENTRANCE(SWORDMANS_SCHOOL, 0),             ENTRANCE(WEST_CLOCK_TOWN, 3),
+                 DAY(1) || DAY(2) || (DAY(3) && IS_DAY && TIME_AT_LEAST(6, 0) && TIME_BEFORE(23, 0)) || NIGHT(3)
```
Note to self: return to this. Order of operations does not look right.

Eblo commented on mm/2s2h/Rando/Logic/Regions/North.cpp:
```diff
@@ -368,7 +368,7 @@ static RegisterShipInitFunc initFunc([]() {
         },
         .events = {
-            EVENT(RE_ACCESS_PICTOGRAPH_TINGLE, HAS_ITEM(ITEM_PICTOGRAPH_BOX)),
+            EVENT(RE_ACCESS_PICTOGRAPH_TINGLE, IS_DAY && HAS_ITEM(ITEM_PICTOGRAPH_BOX)),
```
Tingle appears at all times here. I believe North Clock Town is the only Tingle location that is time dependent.
```suggestion
            EVENT(RE_ACCESS_PICTOGRAPH_TINGLE, HAS_ITEM(ITEM_PICTOGRAPH_BOX)),
```

Eblo commented on mm/2s2h/Rando/Logic/Regions/South.cpp:
```diff
@@ -209,13 +209,17 @@ static RegisterShipInitFunc initFunc([]() {
         },
         .events = {
-            EVENT(RE_ACCESS_PICTOGRAPH_TINGLE, HAS_ITEM(ITEM_PICTOGRAPH_BOX)),
+            EVENT(RE_ACCESS_PICTOGRAPH_TINGLE, IS_DAY && HAS_ITEM(ITEM_PICTOGRAPH_BOX)),
```
Ditto.
```suggestion
            EVENT(RE_ACCESS_PICTOGRAPH_TINGLE, HAS_ITEM(ITEM_PICTOGRAPH_BOX)),
```

Eblo commented on mm/2s2h/Rando/Logic/Regions/West.cpp:
```diff
@@ -162,7 +165,7 @@ static RegisterShipInitFunc initFunc([]() {
         },
         .events = {
-            EVENT(RE_ACCESS_PICTOGRAPH_TINGLE, HAS_ITEM(ITEM_PICTOGRAPH_BOX)),
+            EVENT(RE_ACCESS_PICTOGRAPH_TINGLE, IS_DAY && HAS_ITEM(ITEM_PICTOGRAPH_BOX)),
```
Ditto.
```suggestion
            EVENT(RE_ACCESS_PICTOGRAPH_TINGLE, HAS_ITEM(ITEM_PICTOGRAPH_BOX)),
```

Eblo commented on mm/2s2h/Rando/Logic/Logic.h:
```diff
@@ -98,6 +98,45 @@ extern std::unordered_map<RandoRegionId, RandoRegion> Regions;
+                                                                                      : false)))
+// Coarse hour windows map to day/night halves for logic purposes
+#define TIME_AT_LEAST(h, m) ((h) >= 18 ? IS_NIGHT : IS_DAY)
+#define TIME_BEFORE(h, m) ((h) < 6 ? IS_NIGHT : IS_DAY)
```
Many uses of these macros are ANDed with `IS_NIGHT` or `IS_DAY`, which makes this condition meaningless. Look at `RC_TERMINA_FIELD_KAMARO_MASK` for example:

`IS_NIGHT && TIME_AT_LEAST(0, 0) && TIME_BEFORE(5, 0)`
expands to
`IS_NIGHT && ((0) >= 18 ? IS_NIGHT : IS_DAY) && ((5) < 6 ? IS_NIGHT : IS_DAY)`
which becomes
`IS_NIGHT && IS_DAY && IS_NIGHT`

Which is not correct in this case, as only night matters for that check. The deeper question is whether these macros are needed at all. The player can only acquire 12 hour chunks of time, so logic only has those 12 hour chunks to deal with. I feel like these macros add complexity where it is not needed.

Eblo commented on mm/2s2h/Rando/MiscBehavior/ClockShuffle.cpp:
```diff
@@ -0,0 +1,592 @@
+        case RI_CLOCK_NIGHT_3:
+            return HALF_DAY3_NIGHT;
+        default:
+            return -1; // Invalid item
```
Might be good to have a `ClockHalfIndex` enum value equating to -1.

Eblo commented on mm/2s2h/Rando/MiscBehavior/ClockShuffle.cpp:
```diff
@@ -0,0 +1,592 @@
+    });
+
+    // Force a scene transition to apply the time change
+    ForceSceneReload();
```
This causes reloads even in scenes that do not normally do so with day transitions (e.g. dungeons). This can be a tad annoying. Scenes of this nature do not seem to have any time-sensitive states, so skipping this behavior in those scenes should be safe. The non-intrusive banner message should be enough telegraphing to the player that time has changed.

Eblo commented on mm/2s2h/Rando/MiscBehavior/ClockShuffle.cpp:
```diff
@@ -0,0 +1,592 @@
+    COND_HOOK(OnSceneInit, IS_RANDO && RANDO_SAVE_OPTIONS[RO_CLOCK_SHUFFLE], [](s16 sceneId, s8 spawnNum) {
+        // Handle pending DayTelop, but delay if player is in Clock Tower Interior
+        if (sPendingDayTelopDay != -1 && gPlayState != nullptr) {
+            // Don't trigger DayTelop if we're in Clock Tower Interior (narratively inconsistent)
```
Is this even possible? Time and SoDT are not functional there. Ditto other scenes like the moon.

mckinlee replied:
It is. Because we do a time correction after time is set to day 0 0559, it sets up a pending daytelop which would then trigger inside the clock tower without this. This happens because the time change to day 0 happens in the previous scene.

Eblo commented on mm/2s2h/Rando/MiscBehavior/ClockShuffle.cpp:
```diff
@@ -0,0 +1,592 @@
+    COND_HOOK(OnPlayDestroy, IS_RANDO && RANDO_SAVE_OPTIONS[RO_CLOCK_SHUFFLE], []() {
+        if (gSaveContext.save.day == 0 && gSaveContext.save.time == DAY_0_0559_TIME) {
+            if (Flags_GetRandoInf(RANDO_INF_OBTAINED_CLOCK_DAY_1)) {
+                // Player owns Day 1, allow natural progression
```
Since this condition has no behavior, I'd prefer to just invert it and drop the else. The comment can remain to explain the condition.

Eblo commented on mm/2s2h/Rando/StaticData/Items.cpp:
```diff
@@ -176,6 +176,13 @@ std::map<RandoItemId, RandoStaticItem> Items = {
+    RI(RI_CLOCK_NIGHT_2,              "a",     "Clock (Night 2)",            RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_CLOCK_DAY_3,                "a",     "Clock (Day 3)",              RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_CLOCK_NIGHT_3,              "a",     "Clock (Night 3)",            RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_CLOCK_PROGRESSIVE,          "a",     "Progressive Clock",          RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
```
```suggestion
    RI(RI_CLOCK_DAY_1,                "a",    "Clock (Day 1)",              RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
    RI(RI_CLOCK_NIGHT_1,              "a",    "Clock (Night 1)",            RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
    RI(RI_CLOCK_DAY_2,                "a",    "Clock (Day 2)",              RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
    RI(RI_CLOCK_NIGHT_2,              "a",    "Clock (Night 2)",            RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
    RI(RI_CLOCK_DAY_3,                "a",    "Clock (Day 3)",              RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
    RI(RI_CLOCK_NIGHT_3,              "a",    "Clock (Night 3)",            RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
    RI(RI_CLOCK_PROGRESSIVE,          "a",    "Progressive Clock",          RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
```

Eblo commented on mm/2s2h/Rando/StaticData/Options.cpp:
```diff
@@ -47,6 +47,9 @@ std::map<RandoOptionId, RandoStaticOption> Options = {
     RO(RO_STARTING_RUPEES,             RO_GENERIC_OFF),
+    RO(RO_CLOCK_SHUFFLE,             RO_GENERIC_OFF),
+    RO(RO_CLOCK_SHUFFLE_PROGRESSIVE,     RO_CLOCK_SHUFFLE_RANDOM),
+    RO(RO_CLOCK_TERMINAL_TIME,       0), // Default: 00:00 (midnight)
```
```suggestion
    RO(RO_CLOCK_SHUFFLE,               RO_GENERIC_OFF),
    RO(RO_CLOCK_SHUFFLE_PROGRESSIVE,   RO_CLOCK_SHUFFLE_RANDOM),
    RO(RO_CLOCK_TERMINAL_TIME,         0), // Default: 00:00 (midnight)
```

Eblo commented on mm/2s2h/Rando/DrawFuncs.cpp:
```diff
@@ -22,7 +24,13 @@ s32 EnMinifrog_OverrideLimbDraw(PlayState* play, s32 limbIndex, Gfx** dList, Vec
 /* Minifrog */  #include "objects/object_fr/object_fr.h"
-// clang-format on
+/* Clock */     #include "overlays/actors/ovl_Obj_Tokeidai/z_obj_tokeidai.h"
+    // clang-format on
```
I think the spaces here caused clang to miss the tag and not format the rest of the file.
```suggestion
// clang-format on
```

mckinlee replied:
Weird. I did that and the script still formatted it strangely. Not sure whats up with that, but we could move  // clang-format on to be beneath ObjTokeidai_RotateOnHourChange if you agree.

Eblo commented on mm/2s2h/Rando/DrawFuncs.cpp:
```diff
@@ -261,4 +269,80 @@ extern void DrawMinifrog(RandoItemId randoItemId, Actor* actor) {
+            sunMoonPanelRotation = 0x8000;
+            break;
+        case RI_CLOCK_PROGRESSIVE:
+            clockFaceRotation = gSaveContext.save.isNight ? 0 : 0xC000;
+            sunMoonPanelRotation = gSaveContext.save.isNight ? 0x8000 : 0;
```
Do these values come from anywhere in the code, or is this just from observing in real-time?

mckinlee replied:
The sunMoonPanelRotation is based off ObjTokeidai_GetTargetSunMoonPanelRotation but clockFaceRotation was based on my own observations.. it's empirically determined because the vanilla formula (GET_CLOCK_FACE_ROTATION) is context-specific to the fixed Clock Tower building, and doesn't apply to the scaled-down floating pickup model.

Eblo commented on mm/2s2h/Rando/Menu.cpp:
```diff
@@ -201,6 +205,48 @@ static void DrawLocationsTab() {
     CVarCheckbox("Shuffle Owl Statues", Rando::StaticData::Options[RO_SHUFFLE_OWL_STATUES].cvar);
     CVarCheckbox("Shuffle Shops", Rando::StaticData::Options[RO_SHUFFLE_SHOPS].cvar);
     CVarCheckbox("Shuffle Tingle Maps", Rando::StaticData::Options[RO_SHUFFLE_TINGLE_SHOPS].cvar);
+    CVarCheckbox("Shuffle Clock", Rando::StaticData::Options[RO_CLOCK_SHUFFLE].cvar,
```
As discussed, this better belongs in the Items tab.

Eblo commented on mm/2s2h/Rando/StaticData/Checks.cpp:
```diff
@@ -2228,6 +2228,15 @@ std::map<RandoCheckId, RandoStaticCheck> Checks = {
+    RC(RC_CLOCK_NIGHT_2,                                 RCTYPE_CLOCK, SCENE_MAX,                FLAG_NONE, 0x00,                                                RI_CLOCK_NIGHT_2),
+    RC(RC_CLOCK_DAY_3,                                   RCTYPE_CLOCK, SCENE_MAX,                FLAG_NONE, 0x00,                                                RI_CLOCK_DAY_3),
+    RC(RC_CLOCK_NIGHT_3,                                 RCTYPE_CLOCK, SCENE_MAX,                FLAG_NONE, 0x00,                                                RI_CLOCK_NIGHT_3),
+    RC(RC_CLOCK_PROGRESSIVE,                             RCTYPE_CLOCK, SCENE_MAX,                FLAG_NONE, 0x00,                                                RI_CLOCK_PROGRESSIVE),
```
```suggestion
    RC(RC_CLOCK_DAY_1,                                   RCTYPE_CLOCK, SCENE_MAX,               FLAG_NONE, 0x0, RI_CLOCK_DAY_1),
    RC(RC_CLOCK_NIGHT_1,                                 RCTYPE_CLOCK, SCENE_MAX,               FLAG_NONE, 0x0, RI_CLOCK_NIGHT_1),
    RC(RC_CLOCK_DAY_2,                                   RCTYPE_CLOCK, SCENE_MAX,               FLAG_NONE, 0x0, RI_CLOCK_DAY_2),
    RC(RC_CLOCK_NIGHT_2,                                 RCTYPE_CLOCK, SCENE_MAX,               FLAG_NONE, 0x0, RI_CLOCK_NIGHT_2),
    RC(RC_CLOCK_DAY_3,                                   RCTYPE_CLOCK, SCENE_MAX,               FLAG_NONE, 0x0, RI_CLOCK_DAY_3),
    RC(RC_CLOCK_NIGHT_3,                                 RCTYPE_CLOCK, SCENE_MAX,               FLAG_NONE, 0x0, RI_CLOCK_NIGHT_3),
    RC(RC_CLOCK_PROGRESSIVE,                             RCTYPE_CLOCK, SCENE_MAX,               FLAG_NONE, 0x0, RI_CLOCK_PROGRESSIVE),
```

Eblo commented on mm/2s2h/Rando/Logic/Regions/Central.cpp:
This chest only appears once the tower is fully built.

Eblo commented on mm/2s2h/Rando/MiscBehavior/ClockShuffle.cpp:
```diff
@@ -0,0 +1,592 @@
+    COND_VB_SHOULD(VB_SUPPRESS_DAY_INCREMENT, IS_RANDO && RANDO_SAVE_OPTIONS[RO_CLOCK_SHUFFLE] && sPendingDayTelopDay,
+                   { *should = true; });
+
+    // Hook all Song of Double Time message IDs with a single handler
```
The regular Song of Time should be overridden as well, as now it goes to the earliest available time slot rather than Dawn of the First Day. The message ID is `0x1b8a`.

Eblo commented on mm/2s2h/Rando/Logic/Regions/Central.cpp:
This is gated by night 3.

Eblo commented on mm/2s2h/Rando/Logic/Regions/Central.cpp:
```diff
@@ -299,20 +405,22 @@ static RegisterShipInitFunc initFunc([]() {
         .checks = {
-            CHECK(RC_DEKU_PLAYGROUND_ALL_DAYS, CAN_BE_DEKU),
-            CHECK(RC_DEKU_PLAYGROUND_ANY_DAY, CAN_BE_DEKU),
+            CHECK(RC_DEKU_PLAYGROUND_ALL_DAYS, CAN_BE_DEKU && TIME_AT_LEAST(6, 0)),
```
This is now going to require one part of each individual day.

Eblo commented on mm/2s2h/Rando/Logic/Regions/Central.cpp:
```diff
@@ -299,20 +405,22 @@ static RegisterShipInitFunc initFunc([]() {
-            CHECK(RC_DEKU_PLAYGROUND_ALL_DAYS, CAN_BE_DEKU),
-            CHECK(RC_DEKU_PLAYGROUND_ANY_DAY, CAN_BE_DEKU),
+            CHECK(RC_DEKU_PLAYGROUND_ALL_DAYS, CAN_BE_DEKU && TIME_AT_LEAST(6, 0)),
+            CHECK(RC_DEKU_PLAYGROUND_ANY_DAY, CAN_BE_DEKU && TIME_AT_LEAST(6, 0)),
```
Does there need to be any time condition here? The playground is always open, isn't it?

Eblo commented on mm/2s2h/Rando/Logic/Regions/MilkRoad.cpp:
```diff
@@ -93,12 +98,12 @@ static RegisterShipInitFunc initFunc([]() {
         .exits = { //     TO                                         FROM
             EXIT(ENTRANCE(TERMINA_FIELD, 5),                ENTRANCE(MILK_ROAD, 0), true),
-            EXIT(ENTRANCE(ROMANI_RANCH, 0),                 ENTRANCE(MILK_ROAD, 1), true),
+            EXIT(ENTRANCE(ROMANI_RANCH, 0),                 ENTRANCE(MILK_ROAD, 1), (DAY(3) && TIME_AT_LEAST(6, 0)) || HAS_ITEM(ITEM_POWDER_KEG)),
```
Day or night 3, though this change is pending my questions about accounting for clock hours.

Eblo commented on mm/2s2h/Rando/Logic/Regions/MilkRoad.cpp:
```diff
@@ -93,12 +98,12 @@ static RegisterShipInitFunc initFunc([]() {
             EXIT(ENTRANCE(TERMINA_FIELD, 5),                ENTRANCE(MILK_ROAD, 0), true),
-            EXIT(ENTRANCE(ROMANI_RANCH, 0),                 ENTRANCE(MILK_ROAD, 1), true),
+            EXIT(ENTRANCE(ROMANI_RANCH, 0),                 ENTRANCE(MILK_ROAD, 1), (DAY(3) && TIME_AT_LEAST(6, 0)) || HAS_ITEM(ITEM_POWDER_KEG)),
             EXIT(ENTRANCE(GORMAN_TRACK, 3),                 ENTRANCE(MILK_ROAD, 2), RANDO_EVENTS[RE_COWS_FROM_ALIENS]),
```
This is, I believe, night 2 only.

Eblo commented on mm/2s2h/Rando/Logic/Regions/MilkRoad.cpp:
```diff
@@ -93,12 +98,12 @@ static RegisterShipInitFunc initFunc([]() {
-            EXIT(ENTRANCE(ROMANI_RANCH, 0),                 ENTRANCE(MILK_ROAD, 1), true),
+            EXIT(ENTRANCE(ROMANI_RANCH, 0),                 ENTRANCE(MILK_ROAD, 1), (DAY(3) && TIME_AT_LEAST(6, 0)) || HAS_ITEM(ITEM_POWDER_KEG)),
             EXIT(ENTRANCE(GORMAN_TRACK, 3),                 ENTRANCE(MILK_ROAD, 2), RANDO_EVENTS[RE_COWS_FROM_ALIENS]),
             EXIT(ENTRANCE(GORMAN_TRACK, 0),                 ENTRANCE(MILK_ROAD, 3), true),
```
Milk road may need split into new regions. The fencing is different on night 2 if you saved the ranch from aliens. If memory serves, there are other fence layouts as well, but I'm not sure.

Eblo commented on mm/2s2h/Rando/Logic/Regions/MilkRoad.cpp:
```diff
@@ -93,12 +98,12 @@ static RegisterShipInitFunc initFunc([]() {
         },
         .events = {
-            EVENT(RE_ACCESS_PICTOGRAPH_TINGLE, HAS_ITEM(ITEM_PICTOGRAPH_BOX)),
+            EVENT(RE_ACCESS_PICTOGRAPH_TINGLE, IS_DAY && HAS_ITEM(ITEM_PICTOGRAPH_BOX)),
```
```suggestion
            EVENT(RE_ACCESS_PICTOGRAPH_TINGLE, HAS_ITEM(ITEM_PICTOGRAPH_BOX)),
```

Eblo commented on mm/2s2h/Rando/Logic/Regions/MilkRoad.cpp:
```diff
@@ -121,13 +128,13 @@ static RegisterShipInitFunc initFunc([]() {
     Regions[RR_ROMANI_RANCH] = RandoRegion{ .sceneId = SCENE_F01,
         .checks = {
-            CHECK(RC_ROMANI_RANCH_ALIENS, HAS_ITEM(ITEM_BOW) && CAN_BE_GORON && HAS_ITEM(ITEM_POWDER_KEG)),
+            CHECK(RC_ROMANI_RANCH_ALIENS, (NIGHT(1) && TIME_AT_LEAST(2, 30)) && HAS_ITEM(ITEM_BOW)),
```
Wow, all these checks have Goron and powder keg as requirements. The entrance already takes care of that. That's a note for cleanup beyond this PR.

Eblo commented on mm/2s2h/Rando/Logic/Regions/East.cpp:
```diff
@@ -177,7 +177,7 @@ static RegisterShipInitFunc initFunc([]() {
         },
         .events = {
-            EVENT(RE_ACCESS_PICTOGRAPH_TINGLE, HAS_ITEM(ITEM_PICTOGRAPH_BOX)),
+            EVENT(RE_ACCESS_PICTOGRAPH_TINGLE, IS_DAY && HAS_ITEM(ITEM_PICTOGRAPH_BOX)),
```
```suggestion
            EVENT(RE_ACCESS_PICTOGRAPH_TINGLE, HAS_ITEM(ITEM_PICTOGRAPH_BOX)),
```

Eblo reviewed (changes requested):
First wave of feedback with plenty of questions. I think you know this already, but it looks like there is strangely a crash in `CollisionCheck_SetAC` for the clocks. This only seems to happen for the Zora-rangable ones, which call that function. I'm not sure what that's about. Maybe the weirdness with how the animated drawfunc is handled causes it to mess with its collider?

> BetterSoDT feedback.. might leave this for a potential v2.. can't come up with a good solution atm

Current behavior is to load the time you selected, then abort and load the next available time if that isn't one. Not great. I think the most painless route is to just check if the selected time is actually available, and if not, tell the player as much instead of charging ahead.

All that aside, while I didn't do a full run, the bits I did were a fairly interesting way of challenging my understanding of this schedule-driven game. I think this will be a fun shuffle for sure.

mckinlee committed:
logic refactor, minor changes, and new enhancements

mckinlee committed:
Merge remote-tracking branch 'upstream/develop' into clocks-as-items

mckinlee committed:
huge rework

mckinlee committed:
Merge remote-tracking branch 'upstream/develop' into clocks-as-items

mckinlee committed:
clang

mckinlee committed:
fix a oops; any day, not current day

mckinlee committed:
housekeeping and starting items menu support

mckinlee committed:
Merge remote-tracking branch 'upstream/develop' into clocks-as-items

mckinlee committed:
splits up some milkroad regions and fixes a audio bug

mckinlee committed:
Merge remote-tracking branch 'upstream/develop' into clocks-as-items

mckinlee committed:
i suck at merging

mckinlee committed:
Merge remote-tracking branch 'upstream/develop' into clocks-as-items

Eblo commented on mm/2s2h/Rando/Logic/Regions/MilkRoad.cpp:
```diff
@@ -121,13 +137,13 @@ static RegisterShipInitFunc initFunc([]() {
+            CHECK(RC_ROMANI_RANCH_EPONAS_SONG, (HAS_ITEM(ITEM_POWDER_KEG) && CAN_BE_GORON) && BEFORE(TIME_NIGHT1_PM_06_00)),
+            CHECK(RC_ROMANI_RANCH_FIELD_COW_ENTRANCE, CAN_PLAY_SONG(EPONA)),
+            CHECK(RC_ROMANI_RANCH_FIELD_COW_NEAR_HOUSE_BACK, CAN_PLAY_SONG(EPONA)),
+            CHECK(RC_ROMANI_RANCH_FIELD_COW_NEAR_HOUSE_FRONT, CAN_PLAY_SONG(EPONA)),
```
Like the barn cows, these are not accessible if They are not stopped. The times may be different, however, as I think in-universe they are the same as the barn cows.

Eblo commented on mm/2s2h/Rando/Logic/Regions/North.cpp:
```diff
@@ -336,19 +336,19 @@ static RegisterShipInitFunc initFunc([]() {
             CHECK(RC_TWIN_ISLANDS_LARGE_SNOWBALL_09, CanKillEnemy(ACTOR_OBJ_SNOWBALL)),
             CHECK(RC_TWIN_ISLANDS_LARGE_SNOWBALL_10, CanKillEnemy(ACTOR_OBJ_SNOWBALL)),
             CHECK(RC_TWIN_ISLANDS_LARGE_SNOWBALL_11, CanKillEnemy(ACTOR_OBJ_SNOWBALL)),
             CHECK(RC_TWIN_ISLANDS_LARGE_SNOWBALL_12, CanKillEnemy(ACTOR_OBJ_SNOWBALL)),
```
This snowball, for no apparent reason, does not exist on day or night 2.

Eblo commented on mm/2s2h/Rando/Logic/Regions/South.cpp:
```diff
@@ -488,21 +488,21 @@ static RegisterShipInitFunc initFunc([]() {
-            CHECK(RC_WOODS_OF_MYSTERY_GRASS_09, true),
-            CHECK(RC_WOODS_OF_MYSTERY_GRASS_10, true),
+            CHECK(RC_WOODS_OF_MYSTERY_GRASS_09, FINAL_DAY()),
+            CHECK(RC_WOODS_OF_MYSTERY_GRASS_10, FINAL_DAY()),
```
It'll be dependent on the PR being merged, but there are two more grass checks that are only available on day 3: https://github.com/HarbourMasters/2ship2harkinian/pull/1297/files#diff-5a2e7091b84fe5f20ebba96786d781f8f99f614442979320e75c8d87deb3362dR2203-R2204

Eblo commented on mm/2s2h/Rando/MiscBehavior/ClockShuffle.cpp:
```diff
@@ -0,0 +1,733 @@
+            } else {
+                // Handle wrap-around (e.g., 23:00 to 06:00)
+                timeDiff = (DAWN_TIME + CLOCK_TIME(24, 0)) - terminalTime;
+            }
```
I wonder if this would be better served by using `ZERO_DAY_START`. That's currently defined in Traps.cpp, but it seems like it could get some common use. If you agree, it could be moved to ShipUtils or somewhere similar.

Eblo commented on mm/2s2h/Rando/MiscBehavior/ClockShuffle.cpp:
```diff
@@ -0,0 +1,733 @@
+                // Use different message format for each song
+                if (isSongOfTime) {
+                    entry.msg = "Save and return to " + destinationText + "?\n%gYes\nNo\xC2";
+                } else {
```
```suggestion
                } else { // Song of Double Time
```

Eblo reviewed (changes requested):
With enemy drops now being merged, this will now need to account for enemies that are exclusive to certain times. My review on that PR should have you covered for a primer: https://github.com/HarbourMasters/2ship2harkinian/pull/1282/files#diff-69c6aa4d2342ae40aead6a9a0efd7dee2c7ed6788bf2f998c60e36fb769f1e7f

This will also need rebased on the latest develop.

mckinlee committed:
Merge branch 'main' into clocks-as-items

mckinlee committed:
feedback update

Eblo commented:
Quick thought that just occurred to me is that the scarecrow is another way in the game to advance time. I just tried it in a seed where I had Night 2 and Night 3. Tried it on Night 2, which advanced to Day 3, which then panicked and moon crashed. I left the clock tower to see Dawn of the First Day, which then loaded Night 2.

I wouldn't want to lose the scarecrow behavior, so maybe this can be patched to just advance to the next chunk, or imitate vanilla Night 3 behavior if you're already on your latest chunk.

And now that I think about it, granny's stories are another thing to test.

Edit: Upon testing, granny seems fine at least.

mckinlee committed:
scarecrow dance fix

mckinlee committed:
Merge branch 'main' into clocks-as-items

mckinlee committed:
add clocks to new item tracker

mckinlee committed:
day 2 bean patches

Eblo commented on mm/2s2h/Rando/Logic/Logic.h:
```diff
@@ -210,6 +210,12 @@ void ValidateRegionTimeOwnership(RandoRegionId regionId, RandoCheckId checkId, u
+// Requires Day 2 access OR manual watering capability
+#define CAN_USE_DAY2_RAIN_BEAN                                               \
+    (HAS_ITEM(ITEM_MAGIC_BEANS) && (CLOCK_DAY2() || CAN_PLAY_SONG(STORMS) || \
+                                    (HAS_BOTTLE && (CAN_ACCESS(SPRING_WATER) || CAN_ACCESS(HOT_SPRING_WATER)))))
```
Probably simpler to define this as CAN_GROW_BEAN_PLANT or day 2 and bean:

```suggestion
#define CAN_USE_DAY2_RAIN_BEAN (CAN_GROW_BEAN_PLANT || (HAS_ITEM(ITEM_MAGIC_BEANS) && CLOCK_DAY2()))
```

Eblo reviewed (commented):
Small feedback update. One last thing is that my recent enemy drops v2 PR changes how enemy drops are defined. I went ahead and commented any sort of time window logic for your convenience. I'll want to do a more thorough review of this after that enemy drops PR is merged and incorporated into this one, but I found no issues after the scarecrow fix. Clock Shuffle (name pending?) is looking to be in pretty good shape.

mckinlee commented:
> Small feedback update. One last thing is that my recent enemy drops v2 PR changes how enemy drops are defined. I went ahead and commented any sort of time window logic for your convenience. I'll want to do a more thorough review of this after that enemy drops PR is merged and incorporated into this one, but I found no issues after the scarecrow fix. Clock Shuffle (name pending?) is looking to be in pretty good shape.

When that happens, I'll go back and rebrand to "Shuffle Time" as discussed in discord.

Eblo commented:
With recent merges, this will need to do the following with a rebase:

- Move new VBs to `GameInteractor_VanillaBehavior.h` and document their args and vanilla conditions
- Account for enemy drops rework with the embedded checks. I took the liberty of commenting any time windows I caught in my PR, so if I didn't miss anything, it should be as simple as replacing my comments with your appropriate time macros: https://github.com/HarbourMasters/2ship2harkinian/pull/1373/changes#diff-f313e2dd472a45ee68a872384bf0b2aba4188d6b0e0e20e2a0fb9c64b1cef5c1

mckinlee committed:
Merge branch 'main' into clocks-as-items

mckinlee committed:
add time logic to enemy drops and simplify day 2 bean macro

mckinlee committed:
rebrand to "Shuffle Time"

Eblo commented on mm/2s2h/Rando/Logic/Regions/North.cpp:
```diff
@@ -269,8 +269,8 @@ static RegisterShipInitFunc initFunc([]() {
-            CHECK(RC_ENEMY_DROP_TEKTITE, CanKillEnemy(ACTOR_OBJ_SNOWBALL) && CanKillEnemy(ACTOR_EN_TITE)), // Day 1 and 3 only
-            CHECK(RC_ENEMY_DROP_WOLFOS, CanKillEnemy(ACTOR_OBJ_SNOWBALL) && CanKillEnemy(ACTOR_EN_WF)), //  Day 2 only
+            CHECK(RC_ENEMY_DROP_TEKTITE, CanKillEnemy(ACTOR_OBJ_SNOWBALL) && CanKillEnemy(ACTOR_EN_TITE) && (IS_DAY1() || IS_DAY3())), // Day 1 and 3 only
+            CHECK(RC_ENEMY_DROP_WOLFOS, CanKillEnemy(ACTOR_OBJ_SNOWBALL) && CanKillEnemy(ACTOR_EN_WF) && IS_DAY2()), // Day 2 only
```
My comments may have been ambiguous. These enemies appear at night also, e.g. Day+Night 2 for the Wolfos:

```suggestion
            CHECK(RC_ENEMY_DROP_TEKTITE, CanKillEnemy(ACTOR_OBJ_SNOWBALL) && CanKillEnemy(ACTOR_EN_TITE) && (FIRST_DAY() || THIRD_DAY())), // Day 1 and 3 only
            CHECK(RC_ENEMY_DROP_WOLFOS, CanKillEnemy(ACTOR_OBJ_SNOWBALL) && CanKillEnemy(ACTOR_EN_WF) && SECOND_DAY()), // Day 2 only
```

Eblo commented on mm/2s2h/Rando/Logic/Regions/North.cpp:
```diff
@@ -394,9 +394,9 @@ static RegisterShipInitFunc initFunc([]() {
-            CHECK(RC_ENEMY_DROP_SNAPPER, CanKillEnemy(ACTOR_OBJ_SNOWBALL) && CanKillEnemy(ACTOR_EN_KAME)), // Day 3 only
+            CHECK(RC_ENEMY_DROP_BOE, CanKillEnemy(ACTOR_EN_MKK) && IS_NIGHT()), // Night only
+            CHECK(RC_ENEMY_DROP_WOLFOS, CanKillEnemy(ACTOR_OBJ_SNOWBALL) && CanKillEnemy(ACTOR_EN_WF) && IS_DAY2()), // Day 2 only
+            CHECK(RC_ENEMY_DROP_SNAPPER, CanKillEnemy(ACTOR_OBJ_SNOWBALL) && CanKillEnemy(ACTOR_EN_KAME) && IS_DAY3()), // Day 3 only
```
```suggestion
            CHECK(RC_ENEMY_DROP_WOLFOS, CanKillEnemy(ACTOR_OBJ_SNOWBALL) && CanKillEnemy(ACTOR_EN_WF) && SECOND_DAY()), // Day 2 only
            CHECK(RC_ENEMY_DROP_SNAPPER, CanKillEnemy(ACTOR_OBJ_SNOWBALL) && CanKillEnemy(ACTOR_EN_KAME) && THIRD_DAY()), // Day 3 only
```

Eblo commented on mm/2s2h/Rando/Logic/Regions/North.cpp:
```diff
@@ -415,9 +415,9 @@ static RegisterShipInitFunc initFunc([]() {
-            CHECK(RC_ENEMY_DROP_SNAPPER, CanKillEnemy(ACTOR_OBJ_SNOWBALL) && CanKillEnemy(ACTOR_EN_KAME)), // Day 3 only
+            CHECK(RC_ENEMY_DROP_BOE, CanKillEnemy(ACTOR_EN_MKK) && IS_NIGHT()), // Night only
+            CHECK(RC_ENEMY_DROP_WOLFOS, CanKillEnemy(ACTOR_OBJ_SNOWBALL) && CanKillEnemy(ACTOR_EN_WF) && IS_DAY2()), // Day 2 only
+            CHECK(RC_ENEMY_DROP_SNAPPER, CanKillEnemy(ACTOR_OBJ_SNOWBALL) && CanKillEnemy(ACTOR_EN_KAME) && IS_DAY3()), // Day 3 only
```
```suggestion
            CHECK(RC_ENEMY_DROP_WOLFOS, CanKillEnemy(ACTOR_OBJ_SNOWBALL) && CanKillEnemy(ACTOR_EN_WF) && SECOND_DAY()), // Day 2 only
            CHECK(RC_ENEMY_DROP_SNAPPER, CanKillEnemy(ACTOR_OBJ_SNOWBALL) && CanKillEnemy(ACTOR_EN_KAME) && THIRD_DAY()), // Day 3 only
```

Eblo commented on mm/2s2h/Rando/Logic/Regions/South.cpp:
Wrong grass checks got marked:

```suggestion
            CHECK(RC_WOODS_OF_MYSTERY_GRASS_22, FINAL_DAY()),
            CHECK(RC_WOODS_OF_MYSTERY_GRASS_23, FINAL_DAY()),
```

Eblo commented on mm/2s2h/Rando/Logic/Regions/South.cpp:
```suggestion
            CHECK(RC_WOODS_OF_MYSTERY_GRASS_09, true),
            CHECK(RC_WOODS_OF_MYSTERY_GRASS_10, true),
```

Eblo commented on mm/2s2h/Enhancements/Trackers/ItemTracker/ItemTrackerSettings.cpp:
```diff
@@ -27,7 +27,7 @@ std::vector<std::string> listOrder = {
 
 std::vector<std::string> randoListOrder = {
-    "Frogs", "Boss Souls", "Owl Statues", "Tingle Maps", "Misc",
+    "Frogs", "Boss Souls", "Owl Statues", "Tingle Maps", "Clocks", "Misc",
```
This is user-facing, so for consistency let's make this "Time"

```suggestion
    "Frogs", "Boss Souls", "Owl Statues", "Tingle Maps", "Time", "Misc",
```

Eblo commented on mm/2s2h/Enhancements/Trackers/ItemTracker/ItemTrackerSettings.cpp:
```diff
@@ -46,6 +46,7 @@ std::map<std::string, std::tuple<int16_t, int16_t, int16_t>> randoItemLists = {
     { "Boss Souls", { RI_SOUL_GOHT, RI_SOUL_TWINMOLD, 5 } },
     { "Owl Statues", { RI_OWL_CLOCK_TOWN_SOUTH, RI_OWL_ZORA_CAPE, 5 } },
     { "Tingle Maps", { RI_TINGLE_MAP_CLOCK_TOWN, RI_TINGLE_MAP_WOODFALL, 6 } },
+    { "Clocks", { RI_CLOCK_DAY_1, RI_CLOCK_NIGHT_3, 6 } },
```
```suggestion
    { "Time", { RI_CLOCK_DAY_1, RI_CLOCK_NIGHT_3, 6 } },
```

Eblo commented on mm/2s2h/Enhancements/Trackers/ItemTracker/ItemTrackerSettings.cpp:
```diff
@@ -249,7 +250,7 @@ void DrawItemList(std::string listName, int columns) {
 
             if (listName == "Frogs" || listName == "Boss Souls" || listName == "Owl Statues" ||
-                listName == "Tingle Maps" || listName == "Misc") {
+                listName == "Tingle Maps" || listName == "Clocks" || listName == "Misc") {
```
```suggestion
                listName == "Tingle Maps" || listName == "Time" || listName == "Misc") {
```

Eblo commented on mm/2s2h/Rando/CheckTracker/CheckTracker.cpp:
```diff
@@ -147,11 +147,15 @@ bool checkTrackerShouldShowRow(bool obtained, bool skipped) {
 
 void CheckTrackerDrawLogicalList() {
```
It seems that CheckTrackerDrawLogicalList is not actually used in 2ship, as its call and condition are commented out. The application functionally only ever calls CheckTrackerDrawNonLogicalList. Given that all of your changes are in this unused function, they effectively do nothing.

How to maintain these two functions is another topic entirely, but in the scope of this PR, I think these changes should be migrated to CheckTrackerDrawNonLogicalList. Performance may be a concern, not sure yet since this new code simply is not running.

mckinlee replied:
Will remove this duplicate/dead logic. CheckTrackerDrawNonLogicalList already calls RefreshChecksInLogic, which already contains the new logic features.

Eblo commented on mm/2s2h/Rando/Logic/Regions/Central.cpp:
```diff
@@ -90,23 +90,24 @@ static RegisterShipInitFunc initFunc([]() {
+            EXIT(ENTRANCE(STOCK_POT_INN, 0),                ENTRANCE(EAST_CLOCK_TOWN, 9), HAS_ITEM(ITEM_ROOM_KEY) || BETWEEN(TIME_DAY1_AM_08_00, TIME_NIGHT1_PM_08_00) || BETWEEN(TIME_DAY2_AM_08_00, TIME_NIGHT2_PM_08_00) || AFTER(TIME_DAY3_AM_08_00)),
             EXIT(ENTRANCE(STOCK_POT_INN, 1),                ENTRANCE(EAST_CLOCK_TOWN, 10), CAN_BE_DEKU), // To upstairs
-            EXIT(ENTRANCE(MILK_BAR, 0),                     ENTRANCE(EAST_CLOCK_TOWN, 11), HAS_ITEM(ITEM_MASK_ROMANI)),
+            EXIT(ENTRANCE(MILK_BAR, 0),                     ENTRANCE(EAST_CLOCK_TOWN, 11), (BETWEEN(TIME_DAY1_AM_10_00, TIME_NIGHT1_PM_09_00) || BETWEEN(TIME_DAY2_AM_10_00, TIME_NIGHT2_PM_09_00) || BETWEEN(TIME_DAY3_AM_10_00, TIME_NIGHT3_PM_09_00)) || (HAS_ITEM(ITEM_MASK_ROMANI) && (BETWEEN(TIME_NIGHT1_PM_10_00, TIME_NIGHT1_AM_05_00) || BETWEEN(TIME_NIGHT2_PM_10_00, TIME_NIGHT2_AM_05_00) || AFTER ...
```
I know we disable clang format for logic, but we need some courtesy line breaks here lol. See if this looks OK in the check tracker:
```suggestion
            EXIT(ENTRANCE(MILK_BAR, 0),                     ENTRANCE(EAST_CLOCK_TOWN, 11), (BETWEEN(TIME_DAY1_AM_10_00, TIME_NIGHT1_PM_09_00) || 
                                                                                            BETWEEN(TIME_DAY2_AM_10_00, TIME_NIGHT2_PM_09_00) || 
                                                                                            BETWEEN(TIME_DAY3_AM_10_00, TIME_NIGHT3_PM_09_00)) || 
                                                                                            (HAS_ITEM(ITEM_MASK_ROMANI) && (BETWEEN(TIME_NIGHT1_PM_10_00, TIME_NIGHT1_AM_05_00) || 
                                                                                            BETWEEN(TIME_NIGHT2_PM_10_00, TIME_NIGHT2_AM_05_00) ||
                                                                                            AFTER(TIME_NIGHT3_PM_10_00)))),
```

Eblo commented on mm/2s2h/Rando/Logic/Regions/Central.cpp:
```diff
@@ -194,61 +219,73 @@ static RegisterShipInitFunc initFunc([]() {
         .checks = {
-            CHECK(RC_CLOCK_TOWN_EAST_HONEY_DARLING_ALL_DAYS, HAS_ITEM(ITEM_BOW) && HAS_ITEM(ITEM_BOMBCHU) && HAS_ITEM(ITEM_BOMB)),
-            CHECK(RC_CLOCK_TOWN_EAST_HONEY_DARLING_ANY_DAY, HAS_ITEM(ITEM_BOW) || HAS_ITEM(ITEM_BOMBCHU) || HAS_ITEM(ITEM_BOMB)),
+            CHECK(RC_CLOCK_TOWN_EAST_HONEY_DARLING_ALL_DAYS, HAS_ITEM(ITEM_BOW) && RANDO_EVENTS[RE_HONEY_DARLING_REWARD_DAY1] && RANDO_EVENTS[RE_HONEY_DARLING_REWARD_DAY2] && RANDO_EVENTS[RE_HONEY_DARLING_REWARD_DAY3]),
```
`HAS_ITEM(ITEM_BOW)` is implicitly covered by the events.
```suggestion
            CHECK(RC_CLOCK_TOWN_EAST_HONEY_DARLING_ALL_DAYS, RANDO_EVENTS[RE_HONEY_DARLING_REWARD_DAY1] && RANDO_EVENTS[RE_HONEY_DARLING_REWARD_DAY2] && RANDO_EVENTS[RE_HONEY_DARLING_REWARD_DAY3]),
```

Eblo commented on mm/2s2h/Rando/Logic/Regions/Central.cpp:
```diff
@@ -194,61 +219,73 @@ static RegisterShipInitFunc initFunc([]() {
+        .events = {
+            EVENT(RE_HONEY_DARLING_REWARD_DAY1, (HAS_ITEM(ITEM_BOMB) || HAS_ITEM(ITEM_BOMBCHU)) && BEFORE(TIME_NIGHT1_PM_10_00)),
+            EVENT(RE_HONEY_DARLING_REWARD_DAY2, HAS_ITEM(ITEM_BOMB) && BETWEEN(TIME_DAY2_AM_06_00, TIME_NIGHT2_PM_10_00)),
+            EVENT(RE_HONEY_DARLING_REWARD_DAY3, (HAS_ITEM(ITEM_BOW) || (CAN_BE_DEKU && HAS_MAGIC)) && IS_DAY3()),
```
Honey and Darling hard block you from attempting the final game if you don't have a bow.

```suggestion
            EVENT(RE_HONEY_DARLING_REWARD_DAY3, HAS_ITEM(ITEM_BOW) && IS_DAY3()),
```

Eblo commented on mm/2s2h/Rando/Logic/Logic.h:
```diff
@@ -185,6 +327,203 @@ inline bool MeetsMoonRequirements() {
+// Requires clocks to access later time periods (Day 2+ or Night 2+)
+#define GRANDMA_STORY_2()                                                                 \
+    ((IS_DAY1() && (CLOCK_DAY2() || CLOCK_NIGHT2() || CLOCK_DAY3() || CLOCK_NIGHT3())) || \
+     (IS_DAY2() && (CLOCK_DAY3() || CLOCK_NIGHT3())))
```
I get that these two conditions are each a mouthful, but since they're only used in one place apiece, I think it would be better to just insert these lengthy conditions into the check definitions themselves, with line breaks.

Eblo commented on mm/2s2h/Rando/Logic/Regions/Central.cpp:
```diff
@@ -194,61 +219,73 @@ static RegisterShipInitFunc initFunc([]() {
                 Flags_GetRandoInf(RANDO_INF_OBTAINED_DEED_MOUNTAIN) || Flags_GetRandoInf(RANDO_INF_OBTAINED_DEED_OCEAN) ||
-                Flags_GetRandoInf(RANDO_INF_OBTAINED_LETTER_TO_MAMA) || Flags_GetRandoInf(RANDO_INF_OBTAINED_LETTER_TO_KAFEI)
+                Flags_GetRandoInf(RANDO_INF_OBTAINED_LETTER_TO_MAMA) || Flags_GetRandoInf(RANDO_INF_OBTAINED_LETTER_TO_KAFEI)) && 
+                MIDNIGHT() && (CLOCK_NIGHT3() || Flags_GetRandoInf(RANDO_INF_OBTAINED_ROOM_KEY) || CAN_BE_DEKU)
```
I'm trying to reason out the logical change here. Shouldn't `(CLOCK_NIGHT3() || Flags_GetRandoInf(RANDO_INF_OBTAINED_ROOM_KEY) || CAN_BE_DEKU)` already be implicitly handled by being able to reach this scene? I imagine maybe you thought the player could enter here at day, have access to night, but not have room key or Deku, and thus get kicked out. But in those circumstances, couldn't the player enter normally, exit onto the rooftop, then just wait until midnight and go back in? Is there some other scenario that warrants this guard?

mckinlee replied:
Not only is it redundant, but the timeStayRestrictions need to be inverted to properly require the room key when outside the open hours.

Eblo commented on mm/2s2h/Rando/Logic/Regions/Central.cpp:
```diff
@@ -194,61 +219,73 @@ static RegisterShipInitFunc initFunc([]() {
-            EXIT(ENTRANCE(POST_OFFICE, 0),                  ENTRANCE(WEST_CLOCK_TOWN, 7), true),
-            EXIT(ENTRANCE(LOTTERY_SHOP, 0),                 ENTRANCE(WEST_CLOCK_TOWN, 8), true),
+            EXIT(ENTRANCE(POST_OFFICE, 0),                  ENTRANCE(WEST_CLOCK_TOWN, 7), BETWEEN(TIME_DAY1_PM_03_00, TIME_NIGHT1_AM_12_00) || (Flags_GetRandoInf(RANDO_INF_OBTAINED_LETTER_TO_KAFEI) && BETWEEN(TIME_NIGHT2_PM_06_00, TIME_NIGHT2_AM_12_00)) || IS_NIGHT3()),
+            EXIT(ENTRANCE(LOTTERY_SHOP, 0),                 ENTRANCE(WEST_CLOCK_TOWN, 8), IS_DAY() || (RANDO_EVENTS[RE_PLAY_LOTTERY] && (BEFORE(TIME_NIGHT1_PM_11_00) || BETWEEN(TIME_NIGHT2_PM_06_00, TIME_NIGHT2_PM_11_00) || BETWEEN(TIME_NIGHT3_PM_06_00, TIME_NIGHT3_PM_11_00)))),
```
I don't think this entrance warrants an event. The condition is just time. Whether the player bought a ticket or not does not impact the ability to enter the shop.

Eblo commented on mm/2s2h/Rando/Logic/Regions/Central.cpp:
```diff
@@ -258,89 +295,123 @@ static RegisterShipInitFunc initFunc([]() {
         },
+        .events = {
+            EVENT(RE_PLAY_LOTTERY, IS_DAY()),
+        },
```
See other comment regarding the entrance condition.

Eblo commented on mm/2s2h/Rando/Logic/Regions/Central.cpp:
```diff
@@ -258,89 +295,123 @@ static RegisterShipInitFunc initFunc([]() {
+            EVENT(RE_DEKU_PLAYGROUND_1, FIRST_DAY()),
+            EVENT(RE_DEKU_PLAYGROUND_2, SECOND_DAY()),
+            EVENT(RE_DEKU_PLAYGROUND_3, FINAL_DAY()),
+        },
```
```suggestion
            CHECK(RC_DEKU_PLAYGROUND_DAY_1_RUPEE_01, RANDO_EVENTS[RE_DEKU_PLAYGROUND_1]),
            CHECK(RC_DEKU_PLAYGROUND_DAY_1_RUPEE_02, RANDO_EVENTS[RE_DEKU_PLAYGROUND_1]),
            CHECK(RC_DEKU_PLAYGROUND_DAY_1_RUPEE_03, RANDO_EVENTS[RE_DEKU_PLAYGROUND_1]),
            CHECK(RC_DEKU_PLAYGROUND_DAY_1_RUPEE_04, RANDO_EVENTS[RE_DEKU_PLAYGROUND_1]),
            CHECK(RC_DEKU_PLAYGROUND_DAY_1_RUPEE_05, RANDO_EVENTS[RE_DEKU_PLAYGROUND_1]),
            CHECK(RC_DEKU_PLAYGROUND_DAY_1_RUPEE_06, RANDO_EVENTS[RE_DEKU_PLAYGROUND_1]),
            CHECK(RC_DEKU_PLAYGROUND_DAY_2_RUPEE_01, RANDO_EVENTS[RE_DEKU_PLAYGROUND_2]),
            CHECK(RC_DEKU_PLAYGROUND_DAY_2_RUPEE_02, RANDO_EVENTS[RE_DEKU_PLAYGROUND_2]),
            CHECK(RC_DEKU_PLAYGROUND_DAY_2_RUPEE_03, RANDO_EVENTS[RE_DEKU_PLAYGROUND_2]),
            CHECK(RC_DEKU_PLAYGROUND_DAY_2_RUPEE_04, RANDO_EVENTS[RE_DEKU_PLAYGROUND_2]),
            CHECK(RC_DEKU_PLAYGROUND_DAY_2_RUPEE_05, RANDO_EVENTS[RE_DEKU_PLAYGROUND_2]),
            CHECK(RC_DEKU_PLAYGROUND_DAY_2_RUPEE_06, RANDO_EVENTS[RE_DEKU_PLAYGROUND_2]),
            CHECK(RC_DEKU_PLAYGROUND_DAY_3_RUPEE_01, RANDO_EVENTS[RE_DEKU_PLAYGROUND_3]),
            CHECK(RC_DEKU_PLAYGROUND_DAY_3_RUPEE_02, RANDO_EVENTS[RE_DEKU_PLAYGROUND_3]),
            CHECK(RC_DEKU_PLAYGROUND_DAY_3_RUPEE_03, RANDO_EVENTS[RE_DEKU_PLAYGROUND_3]),
            CHECK(RC_DEKU_PLAYGROUND_DAY_3_RUPEE_04, RANDO_EVENTS[RE_DEKU_PLAYGROUND_3]),
            CHECK(RC_DEKU_PLAYGROUND_DAY_3_RUPEE_05, RANDO_EVENTS[RE_DEKU_PLAYGROUND_3]),
            CHECK(RC_DEKU_PLAYGROUND_DAY_3_RUPEE_06, RANDO_EVENTS[RE_DEKU_PLAYGROUND_3]),
        },
        .exits = { //     TO                                         FROM
            EXIT(ENTRANCE(NORTH_CLOCK_TOWN, 4),             ENTRANCE(DEKU_SCRUB_PLAYGROUND, 0), true),
        },
        .events = {
            EVENT(RE_DEKU_PLAYGROUND_1, CAN_BE_DEKU && FIRST_DAY()),
            EVENT(RE_DEKU_PLAYGROUND_2, CAN_BE_DEKU && SECOND_DAY()),
            EVENT(RE_DEKU_PLAYGROUND_3, CAN_BE_DEKU && FINAL_DAY()),
        },
```

Eblo commented on mm/2s2h/Rando/Logic/Regions/Central.cpp:
```diff
@@ -258,89 +295,123 @@ static RegisterShipInitFunc initFunc([]() {
-            CHECK(RC_CLOCK_TOWN_EAST_SHOOTING_GALLERY_HIGH_SCORE, HAS_ITEM(ITEM_BOW)),
-            CHECK(RC_CLOCK_TOWN_EAST_SHOOTING_GALLERY_PERFECT_SCORE, HAS_ITEM(ITEM_BOW)),
+            CHECK(RC_CLOCK_TOWN_EAST_SHOOTING_GALLERY_HIGH_SCORE, HAS_ITEM(ITEM_BOW) && (BEFORE(TIME_NIGHT1_PM_10_00) || BETWEEN(TIME_DAY2_AM_06_00, TIME_NIGHT2_PM_10_00) || BETWEEN(TIME_DAY3_AM_06_00, TIME_NIGHT3_PM_10_00))),
+            CHECK(RC_CLOCK_TOWN_EAST_SHOOTING_GALLERY_PERFECT_SCORE, HAS_ITEM(ITEM_BOW) && (BEFORE(TIME_NIGHT1_PM_10_00) || BETWEEN(TIME_DAY2_AM_06_00, TIME_NIGHT2_PM_10_00) || BETWEEN(TIME_DAY3_AM_06_00, TIME_NIGHT3_PM_10_00))),
```
It might be simpler and more accurate to gate the entrance itself with the time restriction.

Eblo commented on mm/2s2h/Rando/Logic/Regions/Central.cpp:
```diff
@@ -362,13 +433,21 @@ static RegisterShipInitFunc initFunc([]() {
+            CHECK(RC_CLOCK_TOWN_EAST_TREASURE_CHEST_GAME_DEKU,  CAN_BE_DEKU && (BEFORE(TIME_NIGHT1_PM_10_00) || BETWEEN(TIME_DAY2_AM_06_00, TIME_NIGHT2_PM_10_00) || IS_DAY3())),
+            CHECK(RC_CLOCK_TOWN_EAST_TREASURE_CHEST_GAME_GORON, CAN_BE_GORON && (BEFORE(TIME_NIGHT1_PM_10_00) || BETWEEN(TIME_DAY2_AM_06_00, TIME_NIGHT2_PM_10_00) || IS_DAY3())),
+            CHECK(RC_CLOCK_TOWN_EAST_TREASURE_CHEST_GAME_HUMAN, (BEFORE(TIME_NIGHT1_PM_10_00) || BETWEEN(TIME_DAY2_AM_06_00, TIME_NIGHT2_PM_10_00) || IS_DAY3())), // can be human
+            CHECK(RC_CLOCK_TOWN_EAST_TREASURE_CHEST_GAME_ZORA,  CAN_BE_ZORA && (BEFORE(TIME_NIGHT1_PM_10_00) || BETWEEN(TIME_DAY2_AM_06_00, TIME_NIGHT2_PM_10_00) || IS_DAY3())),
```
Ditto moving the times to the entrance. Unlike Honey and Darling, these other minigames are not dependent on time.

Eblo commented on mm/2s2h/Rando/Logic/Regions/East.cpp:
```diff
@@ -9,7 +9,7 @@ using namespace Rando::Logic;
     Regions[RR_BENEATH_THE_GRAVEYARD_DAMPE] = RandoRegion{ .sceneId = SCENE_DANPEI2TEST,
         .checks = {
-            CHECK(RC_BENEATH_THE_GRAVEYARD_DAMPE_CHEST, HAS_ITEM(ITEM_BOW)),
+            CHECK(RC_BENEATH_THE_GRAVEYARD_DAMPE_CHEST, HAS_ITEM(ITEM_BOW) && IS_NIGHT3()),
```
This scene itself is gated by `IS_NIGHT3()`, so it would be redundant to gate the chest check also.
```suggestion
            CHECK(RC_BENEATH_THE_GRAVEYARD_DAMPE_CHEST, HAS_ITEM(ITEM_BOW)),
```

Eblo commented on mm/2s2h/Rando/Logic/Regions/MilkRoad.cpp:
```diff
@@ -94,22 +99,32 @@ static RegisterShipInitFunc initFunc([]() {
             EXIT(ENTRANCE(GORMAN_TRACK, 0),                 ENTRANCE(MILK_ROAD, 3), true),
         },
+        .connections = {
+            CONNECTION(RR_MILK_ROAD_BEHIND_FENCE, (CAN_BE_GORON && HAS_ITEM(ITEM_BOMB)) || (RANDO_EVENTS[RE_COWS_FROM_ALIENS] && IS_NIGHT2()) || FINAL_DAY()),
```
We only have glitchless logic for now, so Goron bomb shouldn't be considered until we implement glitches/tricks, which is beyond the scope of this PR. Leaving a TODO for the future is fine.
```suggestion
            // TODO: Trick to Goron bomb jump over the fence
            CONNECTION(RR_MILK_ROAD_BEHIND_FENCE, (RANDO_EVENTS[RE_COWS_FROM_ALIENS] && IS_NIGHT2()) || FINAL_DAY()),
```

Eblo commented on mm/2s2h/Rando/Logic/Regions/MilkRoad.cpp:
```diff
@@ -94,22 +99,32 @@ static RegisterShipInitFunc initFunc([]() {
+            EXIT(ENTRANCE(GORMAN_TRACK, 3),                 ENTRANCE(MILK_ROAD, 2), true),
+        },
+        .connections = {
+            CONNECTION(RR_MILK_ROAD, (CAN_BE_GORON && HAS_ITEM(ITEM_BOMB)) || (RANDO_EVENTS[RE_COWS_FROM_ALIENS] && IS_NIGHT2()) || FINAL_DAY()),
```
```suggestion
            // TODO: Trick to Goron bomb jump over the fence
            CONNECTION(RR_MILK_ROAD, (RANDO_EVENTS[RE_COWS_FROM_ALIENS] && IS_NIGHT2()) || FINAL_DAY()),
```

Eblo commented on mm/2s2h/Rando/Logic/Regions/MilkRoad.cpp:
```diff
@@ -123,12 +138,12 @@ static RegisterShipInitFunc initFunc([]() {
+            CHECK(RC_ROMANI_RANCH_EPONAS_SONG, CAN_BE_GORON && HAS_ITEM(ITEM_POWDER_KEG) && BEFORE(TIME_NIGHT1_PM_06_00)),
+            CHECK(RC_ROMANI_RANCH_FIELD_COW_ENTRANCE, (BETWEEN(TIME_NIGHT1_PM_06_00, TIME_NIGHT1_AM_02_30) || (CAN_PLAY_SONG(EPONA) && CAN_BE_GORON && HAS_ITEM(ITEM_POWDER_KEG)))),
+            CHECK(RC_ROMANI_RANCH_FIELD_COW_NEAR_HOUSE_BACK, (BETWEEN(TIME_NIGHT1_PM_06_00, TIME_NIGHT1_AM_02_30) || (CAN_PLAY_SONG(EPONA) && CAN_BE_GORON && HAS_ITEM(ITEM_POWDER_KEG)))),
+            CHECK(RC_ROMANI_RANCH_FIELD_COW_NEAR_HOUSE_FRONT, (BETWEEN(TIME_NIGHT1_PM_06_00, TIME_NIGHT1_AM_02_30) || (CAN_PLAY_SONG(EPONA) && CAN_BE_GORON && HAS_ITEM(ITEM_POWDER_KEG)))),
```
Since accessing this scene before the final day is already gated behind the powder keg, the time condition renders the powder keg conditions moot.
```suggestion
            CHECK(RC_ROMANI_RANCH_EPONAS_SONG, BEFORE(TIME_NIGHT1_PM_06_00)),
            CHECK(RC_ROMANI_RANCH_FIELD_COW_ENTRANCE, (BETWEEN(TIME_NIGHT1_PM_06_00, TIME_NIGHT1_AM_02_30))),
            CHECK(RC_ROMANI_RANCH_FIELD_COW_NEAR_HOUSE_BACK, (BETWEEN(TIME_NIGHT1_PM_06_00, TIME_NIGHT1_AM_02_30))),
            CHECK(RC_ROMANI_RANCH_FIELD_COW_NEAR_HOUSE_FRONT, (BETWEEN(TIME_NIGHT1_PM_06_00, TIME_NIGHT1_AM_02_30))),
```

A cleaner approach might even be to make an event like `RE_DESTROY_MILK_ROAD_BOULDER`.

Eblo commented on mm/2s2h/Rando/Logic/Regions/North.cpp:
```diff
@@ -345,19 +345,19 @@ static RegisterShipInitFunc initFunc([]() {
             CHECK(RC_TWIN_ISLANDS_LARGE_SNOWBALL_11, CanKillEnemy(ACTOR_OBJ_SNOWBALL)),
-            CHECK(RC_TWIN_ISLANDS_LARGE_SNOWBALL_12, CanKillEnemy(ACTOR_OBJ_SNOWBALL)),
-            CHECK(RC_TWIN_ISLANDS_LARGE_SNOWBALL_13, CanKillEnemy(ACTOR_OBJ_SNOWBALL)),
+            CHECK(RC_TWIN_ISLANDS_LARGE_SNOWBALL_12, (FIRST_DAY() || FINAL_DAY()) && CanKillEnemy(ACTOR_OBJ_SNOWBALL)),
```
```suggestion
            CHECK(RC_TWIN_ISLANDS_LARGE_SNOWBALL_12, (FIRST_DAY() || FINAL_DAY()) && CanKillEnemy(ACTOR_OBJ_SNOWBALL)), // Does not exist on Second Day
```

Eblo commented on mm/2s2h/Rando/Logic/Regions/South.cpp:
```diff
@@ -460,8 +459,8 @@ static RegisterShipInitFunc initFunc([]() {
-            CHECK(RC_SWAMP_SHOOTING_GALLERY_HIGH_SCORE, HAS_ITEM(ITEM_BOW)),
-            CHECK(RC_SWAMP_SHOOTING_GALLERY_PERFECT_SCORE, HAS_ITEM(ITEM_BOW)),
+            CHECK(RC_SWAMP_SHOOTING_GALLERY_HIGH_SCORE, HAS_ITEM(ITEM_BOW) && (BEFORE(TIME_NIGHT1_PM_10_00) || BETWEEN(TIME_DAY2_AM_06_00, TIME_NIGHT2_PM_10_00) || BETWEEN(TIME_DAY3_AM_06_00, TIME_NIGHT3_PM_10_00))),
+            CHECK(RC_SWAMP_SHOOTING_GALLERY_PERFECT_SCORE, HAS_ITEM(ITEM_BOW) && (BEFORE(TIME_NIGHT1_PM_10_00) || BETWEEN(TIME_DAY2_AM_06_00, TIME_NIGHT2_PM_10_00) || BETWEEN(TIME_DAY3_AM_06_00, TIME_NIGHT3_PM_10_00))),
```
The entrance itself is gated, so the checks don't need the same restriction.
```suggestion
            CHECK(RC_SWAMP_SHOOTING_GALLERY_HIGH_SCORE, HAS_ITEM(ITEM_BOW)),
            CHECK(RC_SWAMP_SHOOTING_GALLERY_PERFECT_SCORE, HAS_ITEM(ITEM_BOW)),
```

Eblo commented on mm/2s2h/Rando/Logic/Regions/South.cpp:
```diff
@@ -548,21 +547,21 @@ static RegisterShipInitFunc initFunc([]() {
-            CHECK(RC_WOODS_OF_MYSTERY_GRASS_09, true),
-            CHECK(RC_WOODS_OF_MYSTERY_GRASS_10, true),
+            CHECK(RC_WOODS_OF_MYSTERY_GRASS_09, FINAL_DAY()),
+            CHECK(RC_WOODS_OF_MYSTERY_GRASS_10, FINAL_DAY()),
```
GitHub is not letting me comment on unmodified lines, but:

- `RC_WOODS_OF_MYSTERY_GRASS_01`, `RC_WOODS_OF_MYSTERY_GRASS_02`, `RC_WOODS_OF_MYSTERY_GRASS_09` and `RC_WOODS_OF_MYSTERY_GRASS_09` are available on any day
- `RC_WOODS_OF_MYSTERY_GRASS_04` and `RC_WOODS_OF_MYSTERY_GRASS_05` are only on the First Day
- `RC_WOODS_OF_MYSTERY_GRASS_21` is only on the Second Day (same room as grotto)
- `RC_WOODS_OF_MYSTERY_GRASS_22` and `RC_WOODS_OF_MYSTERY_GRASS_23` are only on the Final Day

Eblo commented on mm/2s2h/Rando/Logic/TimeLogic.cpp:
```diff
@@ -0,0 +1,122 @@
+
+// Owned time calculation - aggregates all owned half-day time slices
+uint64_t GetOwnedTimeSlices() {
+    if (!IS_RANDO || !RANDO_SAVE_OPTIONS[RO_CLOCK_SHUFFLE]) {
```
As best as I can tell, this code is only reachable within rando. Is there a need to check `!IS_RANDO` here?

Eblo commented on mm/2s2h/Rando/Logic/TimeLogic.cpp:
```diff
@@ -0,0 +1,122 @@
+        }
+    }
+
+    return timeSlices ? timeSlices : (1ULL << TIME_DAY1_AM_06_00);
```
What does `1ULL << TIME_DAY1_AM_06_00` translate to?

mckinlee replied:
I'll add a comment and switch it out to use TIME_BIT_ONE for clarity. It's just a expression that translates to: 1ULL << 0. Resulting in the value 1. For context, its a bitmask with only the first bit set which represents the specific time slice of Day 1 6AM.

Eblo commented on mm/2s2h/Rando/MiscBehavior/ClockShuffle.cpp:
```diff
@@ -0,0 +1,765 @@
+        case RI_CLOCK_NIGHT_3:
+            return HALF_DAY3_NIGHT;
+        default:
+            return INVALID; // Invalid item
```
You could probably replace this whole thing with some enum arithmetic. Something like:
```suggestion
    return (clockItemId-RI_CLOCK_DAY_1);
```
Which itself might be cleaner to use as a macro rather than a function.

mckinlee replied:
I agree with using the arithmetic approach, but my preference would be to leave it as a function. Easier to debug.

Eblo commented on mm/2s2h/Rando/MiscBehavior/ClockShuffle.cpp:
```diff
@@ -0,0 +1,765 @@
+
+// Check if a specific day/time is owned by the player in ClockShuffle mode
+bool IsTimeOwnedForClockShuffle(s32 day, u16 time) {
+    if (!IS_RANDO || !RANDO_SAVE_OPTIONS[RO_CLOCK_SHUFFLE]) {
```
Ditto re: whether it's necessary to check `!IS_RANDO` in a rando-specific function.

Eblo commented on mm/2s2h/Rando/MiscBehavior/ClockShuffle.cpp:
```diff
@@ -0,0 +1,765 @@
+            // Check if this is a Song of Time or Song of Double Time message
+            bool isSongOfTime = (*textId == 0x1B8A);
+            bool isSongOfDoubleTime =
+                (*textId == 0x1B91 || *textId == 0x1B90 || *textId == 0x1B8F || *textId == 0x1B92 || *textId == 0x1B8E);
```
This hook runs for all text, checking the text ID for conditions later. It might be better to separate most of the function into a helper function, then have 6 distinct OnOpenText hooks with the `isSongOfTime` and `isSongOfDoubleTime` as args.

Eblo commented on mm/2s2h/Rando/MiscBehavior/ClockShuffle.cpp:
```diff
@@ -0,0 +1,765 @@
+    int clockMode = RANDO_SAVE_OPTIONS[RO_CLOCK_SHUFFLE_PROGRESSIVE];
+
+    // Check if player has selected any starting time items
+    std::vector<RandoItemId> startingItems = convertStartingItemsToRandoItemId(RANDO_STARTING_ITEMS, ",");
```
Need a default, in case somehow the player never touched the starting items setting before. clang format may not like this:
```suggestion
    std::vector<RandoItemId> startingItems = convertStartingItemsToRandoItemId(CVarGetString("gRando.StartingItems", RANDO_STARTING_ITEMS_DEFAULT), ",");
```

Eblo commented on mm/2s2h/Rando/MiscBehavior/ClockShuffle.cpp:
```diff
@@ -0,0 +1,765 @@
+        }
+
+        // Add remaining (non-starting) time items to pool
+        // Only works in random mode - progressive mode items are added elsewhere
```
Clarify where that is here, so that anybody working with this knows where to look.

Eblo commented on mm/2s2h/Rando/Types.h:
```diff
@@ -2467,6 +2467,13 @@ typedef enum {
+    RI_CLOCK_NIGHT_2,
+    RI_CLOCK_DAY_3,
+    RI_CLOCK_NIGHT_3,
+    RI_CLOCK_PROGRESSIVE,
```
This list is in alphabetical order for the most part, so these RIs don't belong here.

On another note, the RI names are partially user-facing at least due to the save editor. Should these be renamed to `RI_TIME_*`?

Eblo commented on mm/2s2h/Rando/Logic/Regions/MilkRoad.cpp:
```diff
@@ -94,22 +99,32 @@ static RegisterShipInitFunc initFunc([]() {
-            CHECK(RC_ROMANI_RANCH_BARN_COW_RIGHT, CAN_PLAY_SONG(EPONA) && CAN_BE_GORON && HAS_ITEM(ITEM_POWDER_KEG))
+            CHECK(RC_ROMANI_RANCH_BARN_COW_LEFT, CAN_PLAY_SONG(EPONA) && (BETWEEN(TIME_NIGHT1_PM_06_00, TIME_NIGHT1_AM_02_30) || RANDO_EVENTS[RE_COWS_FROM_ALIENS])),
+            CHECK(RC_ROMANI_RANCH_BARN_COW_MIDDLE, CAN_PLAY_SONG(EPONA) && (BETWEEN(TIME_NIGHT1_PM_06_00, TIME_NIGHT1_AM_02_30) || RANDO_EVENTS[RE_COWS_FROM_ALIENS])),
+            CHECK(RC_ROMANI_RANCH_BARN_COW_RIGHT, CAN_PLAY_SONG(EPONA) && (BETWEEN(TIME_NIGHT1_PM_06_00, TIME_NIGHT1_AM_02_30) || RANDO_EVENTS[RE_COWS_FROM_ALIENS]))
```
The barn, as well as the doggy track and cucco shack, have kickout times. (So does the house, but there are no checks there currently).

Eblo commented on mm/2s2h/Rando/Logic/Regions/MilkRoad.cpp:
```diff
@@ -34,20 +33,20 @@ static RegisterShipInitFunc initFunc([]() {
-            CHECK(RC_GORMAN_MILK_PURCHASE, CAN_AFFORD(RC_GORMAN_MILK_PURCHASE)),
-            CHECK(RC_GORMAN_TRACK_GARO_MASK, CAN_PLAY_SONG(EPONA)),
+            CHECK(RC_GORMAN_MILK_PURCHASE, CAN_AFFORD(RC_GORMAN_MILK_PURCHASE) && IS_DAY()),
+            CHECK(RC_GORMAN_TRACK_GARO_MASK, CAN_PLAY_SONG(EPONA) && CAN_AFFORD(RC_GORMAN_TRACK_GARO_MASK) && IS_DAY()),
```
While the Gorman race does cost money, it's a flat 10 Rupee fee and not tied into a shop check with a price. It's not gated by wallet size.
```suggestion
            CHECK(RC_GORMAN_TRACK_GARO_MASK, CAN_PLAY_SONG(EPONA) && IS_DAY()),
```

Eblo commented on mm/2s2h/Rando/Logic/Regions/Central.cpp:
```diff
@@ -258,89 +295,123 @@ static RegisterShipInitFunc initFunc([]() {
             EVENT(RE_ACCESS_FISH, true),
             EVENT(RE_ACCESS_BUGS, true),
+            EVENT(RE_SETUP_MEET_ANJU, HAS_ITEM(ITEM_MASK_KAFEIS_MASK) && BETWEEN(TIME_DAY1_PM_01_45, TIME_NIGHT1_PM_09_00)),
+            EVENT(RE_MEET_ANJU, RANDO_EVENTS[RE_SETUP_MEET_ANJU] && BETWEEN(TIME_NIGHT1_AM_12_00, TIME_DAY2_AM_06_00) && (Flags_GetRandoInf(RANDO_INF_OBTAINED_ROOM_KEY) || CAN_BE_DEKU)),
```
This seems to be for the midnight meeting, which makes the event name confusing. Maybe just `RE_ANJU_MIDNIGHT_MEETING`?

Eblo reviewed (changes requested):
Top notch job on this so far. In practice in gameplay, this whole thing seems to work splendidly and is just plain fun. My feedback is almost entirely concerned with how things are under the hood, with plenty of questions.

I think that a number of my comments from prior reviews got lost in the mix, as they have not been addressed while others have. I'd like you to carefully go through the entire PR files themselves for review comments. Certain views on GitHub might be only showing so much.

Again, bang up job here. This was no small feat.

mckinlee committed:
Merge branch 'main' into clocks-as-items

mckinlee committed:
more eblo feedback

Eblo commented on mm/2s2h/Rando/StaticData/Items.cpp:
```diff
@@ -181,13 +181,13 @@ std::map<RandoItemId, RandoStaticItem> Items = {
+    RI(RI_TIME_NIGHT_2,               "a",    "Time (Night 2)",             RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_TIME_DAY_3,                 "a",    "Time (Day 3)",               RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_TIME_NIGHT_3,               "a",    "Time (Night 3)",             RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
+    RI(RI_TIME_PROGRESSIVE,           "a",    "Progressive Time",           RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
```
Could probably just not have articles for these items since they're definite.
```suggestion
    RI(RI_TIME_DAY_1,                 "",    "Time (Day 1)",               RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
    RI(RI_TIME_NIGHT_1,               "",    "Time (Night 1)",             RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
    RI(RI_TIME_DAY_2,                 "",    "Time (Day 2)",               RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
    RI(RI_TIME_NIGHT_2,               "",    "Time (Night 2)",             RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
    RI(RI_TIME_DAY_3,                 "",    "Time (Day 3)",               RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
    RI(RI_TIME_NIGHT_3,               "",    "Time (Night 3)",             RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
    RI(RI_TIME_PROGRESSIVE,           "",    "Progressive Time",           RITYPE_MAJOR,           ITEM_NONE,                       GI_NONE,                     GID_NONE),
```

Eblo commented on mm/2s2h/Rando/Types.h:
```diff
@@ -2467,16 +2467,16 @@ typedef enum {
     RI_SWORD_GILDED,
     RI_SWORD_KOKIRI,
     RI_SWORD_RAZOR,
+    RI_TIME_DAY_1,
+    RI_TIME_NIGHT_1,
+    RI_TIME_DAY_2,
+    RI_TIME_NIGHT_2,
+    RI_TIME_DAY_3,
+    RI_TIME_NIGHT_3,
+    RI_TIME_PROGRESSIVE,
```
What do you think of naming the time RIs like so, to keep them both alphabetical and also the same order as what you have? I'm not 100% sure myself:
```suggestion
    RI_TIME_1_DAY,
    RI_TIME_1_NIGHT,
    RI_TIME_2_DAY,
    RI_TIME_2_NIGHT,
    RI_TIME_3_DAY,
    RI_TIME_3_NIGHT,
    RI_TIME_PROGRESSIVE,
```

mckinlee replied:
I'm indifferent, but I left it as is for now. Just let me know how ya wanna proceed. The other feedback is addressed though.

Eblo reviewed (changes requested):
One miscellaneous thing is that my comment regarding `CheckTrackerDrawLogicalList` is still not addressed. That, and there are conflicts with the Central.cpp logic file.

This is really close. This review is trivial things. I'm pretty happy with this overall.

mckinlee committed:
Merge branch 'main' into clocks-as-items

mckinlee committed:
some more feedback

mckinlee committed:
rename feedback

mckinlee committed:
pauseSaveEntrance fix

mckinlee committed:
fix take 2

Eblo committed:
Merge remote-tracking branch 'upstream/develop' into time-shuffle-fixes

Eblo committed:
Time Shuffle cleanup

Fix pause save Time Shuffle bug
Fix event queue daytelop bug
Finish merge commit header include
Clean up with shouldRegister

Co-authored-by: mckinlee <mckinlee@ymail.com>

mckinlee committed:
Merge pull request #15 from Eblo/time-shuffle-cleanup

Time shuffle cleanup
